frontend/public/components/monitoring/receiver-forms/alert-manager-receiver-forms.tsx

```diff
@@ -0,0 +1,414 @@
+  );
+  const matches = _.map(routesOfReceiver?.match || {}, (v, k) => {
+    return { name: k, value: v, isRegex: false };
+  });
+  const regexMatches = _.map(routesOfReceiver?.match_re || {}, (v, k) => {
+    return { name: k, value: v, isRegex: true };
+  });
+  return _.concat([], matches, regexMatches);
+};
+
+/**
+ * Returns new Route object
+ * Ex:
+ * {
+ *   receiver: myNewReceiver,
+ *   match: {
+ *     "severity": "warning",
+ *     "cluster": "myCluster"
+ *   }
+ *   match_re {
+ *     "service": "^(foo1|foo2|baz)$",
+ *   }
+ * }
+ */
+const createRoute = (
+  receiver: AlertmanagerReceiver,
+  routeLabels: RouteEditorLabel[],
+): AlertmanagerRoute => {
+  return _.reduce(
+    routeLabels,
+    (acc, label) => {
+      _.set(acc, [label.isRegex ? 'match_re' : 'match', label.name], label.value);
+      return acc;
+    },
+    { receiver: receiver.name },
+  );
+};
+
+/**
+ * Returns new Receiver object
+ * Ex:
+ * {
+ *   name: MyNewReceiver
+ *   pagerduty_configs: {
+ *     routing_key: <integration_key>
+ *   }
+ * }
+ */
+const createReceiver = (
+  formValues,
+  createReceiverConfig: Function,
+  receiverToEdit: AlertmanagerReceiver,
+): AlertmanagerReceiver => {
+  const receiverConfig = createReceiverConfig(
+    formValues,
+    receiverToEdit && receiverToEdit[formValues.receiverType]
+      ? receiverToEdit[formValues.receiverType][0] // pass in receiver config if editing existing receiver
+      : {},
+  );
+  return {
+    name: formValues.receiverName,
+    [formValues.receiverType]: [{ ...receiverConfig }],
+  };
+};
+
+const subformFactory = (receiverType: string) => {
+  switch (receiverType) {
+    case 'pagerduty_configs':
+      return PagerDutyForm;
+    case 'webhook_configs':
+      return WebhookForm;
+    default:
+      return null;
+  }
+};
+
+function formReducer(formValues, action) {
+  if (action.type === 'setFormValues') {
+    return {
+      ...formValues,
+      ...action.payload,
+    };
+  }
+  // eslint-disable-next-line no-console
+  console.error('Unrecognized Alertmanager form reducer action', action);
+  return formValues;
+}
+
+const getRouteLabelsForEditor = (
+  isDefaultReceiver: boolean,
+  receiverToEdit: AlertmanagerReceiver,
+  allRoutes: AlertmanagerRoute[],
+): RouteEditorLabel[] => {
+  const routeLabels = convertReceiverRoutesToEditorLabels(receiverToEdit, allRoutes);
+  return !isDefaultReceiver && _.isEmpty(routeLabels)
+    ? [{ name: '', value: '', isRegex: false }]
+    : routeLabels;
+};
+
+const ReceiverBaseForm: React.FC<ReceiverBaseFormProps> = ({
+  obj: secret, // Secret "alertmanager-main" which contains alertmanager.yaml config
+  titleVerb,
+  saveButtonText,
+  editReceiverNamed,
+}) => {
+  const [errorMsg, setErrorMsg] = React.useState<string>();
+  const [inProgress, setInProgress] = React.useState<boolean>(false);
+  let config: AlertmanagerConfig;
+  if (!errorMsg) {
+    config = getAlertmanagerConfig(secret, setErrorMsg);
+  }
+
+  const INITIAL_STATE = {
```

Comment:
nit: why all caps in the name?

Reply:
Just so it stands out more as it's being setup in the next couple of code blocks